Third post: to production (#7)
Eomm/fastify-discord-bot-demo: posts/03-to-production.md

it will works as a first step that will be evolved in the next blog post!

To archive this the [`fastify-server-session`](https://github.com/jsumners/fastify-server-session) plugin will
be used: it map a generated SessionId stored on the cookie's browser to a server memory object.
Doing so we can implement the logic:

```
if (the request has not a session) {
  - proceed to authenticate the client with the Discord Authentication Server
  - create a session and store the token
  - view the user profile
} else {
  - view the user profile with the token associated with the request session
}
```

Let's see the code..

The check we need to verify that the user is already logged in should be an [`onRequest` hook](https://www.fastify.io/docs/latest/Hooks/#onrequest):

```js
fastify.addHook('onRequest', function userAlreadyLogged (req, reply, done) {
  if (req.session.token) {
    viewUserProfile(req.session.token, reply)
      .catch(done) // don't forget to manage errors!
    return // do not call `done` to stop the flow
  }
  done()
})
```

Hooks are one of the coolest Fastify's feature!

> The handler when the user logged in will be updated to:

Reviewer: ```suggestion
The user log handler will be updated to:
```

```js
fastify.get('/discord/callback', async function (request, reply) {
  const token = await this.discordOAuth2.getAccessTokenFromAuthorizationCodeFlow(request)
  // server stored: the token object must not be sent to the client
  request.session.token = token
  return viewUserProfile(token, reply)
})
```

And it's done!

## Connect to mongodb with Fastify

Fastify has a plugin for everything!  This time it is the turn of [`fastify-mongodb`](https://github.com/fastify/fastify-mongodb)!

To load it, it is necessary that the configuration is valid.
To do so we are using `fastify-env` already, but the Fastify plugin loading is totally async.
So, we must use [`after()`](https://www.fastify.io/docs/latest/Server/#after) accomplish this task:

```js
import env from 'fastify-env'
// ...
fastify.register(env, { ... })
  .after((err) => {
    if (err) throw err // if the config file has some issue, we must bubble up it
    fastify.register(fastifyMongo, { url: fastify.config.DB_URI })
  })
```

> **WARNING:** right now we will not focus to find the better place where loading the mongodb connection 

Reviewer: ```suggestion
**WARNING:** right now we will not focus on finding the best place to load the mongodb connection 
```

> as said in the _Encapsulation_ paragraph.  We will do this process in future in order to focus on security

Reviewer: ```suggestion
as said in the _Encapsulation_ paragraph.  We will do this process in the future in order to focus on security
```

steps and to learn how to refactor a Fastify application.

Now the `fastify.mongo` decorator has been added to the instance and it will be accessible to
the instance itself and the children contexts.  
At this stage, all the routes can access `fastify.mongo`!

> So the new awesome feature will be to add a page that list all the people that tried our application!

Reviewer: ```suggestion
So the new awesome feature will be to add a page that lists all the people that tried our application!
```

> But this time the page will not be a Server Side Rendering, but the list loading will be via REST API consumption.

Reviewer: ```suggestion
But this time the page will not be rendered through the server, but it will be loaded via REST API consumption.
```

// the great thing is that you may organize these functions wherever you want in the project
// without struggling with tons of `factory-pattern` or clojures!
async function createUser (request, reply) {
  const { userId } = request.params

  await this.mongo.client.db()
    .collection('Users')
    .updateOne(
      { id: userId },
      {
        $set: request.body,
        $push: { visits: new Date() },
        $setOnInsert: { created: new Date() }
      },
      { upsert: true })

  request.log.debug('Track user %s', userId) // fastify has a logger included!
  reply.code(201)
  return { userId }
}

async function searchUsers (request, reply) {
  const { offset, limit } = request.query

  // just a simple pagination
  const query = await this.mongo.client.db().collection('Users')
    .find({}, { projection: { _id: 0, visits: { $slice: -1 } } })
    .sort({ username: 1 })
    .skip(offset)
    .limit(limit)

  const total = await query.count() // it doesn't apply the skip&limit 😉
  const rows = await query.toArray()

  return { rows, total }
}
```

## End

In the next post we will:

+ explore the project structure: is it maintainable?
+ protect the `/api` endpoints

> + migrate from statefull session to a stateless JWT

Reviewer: ```suggestion
+ migrate from stateful session to a stateless JWT
```
